pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java

```diff
@@ -263,52 +263,51 @@ protected void internalDeleteNamespace(AsyncResponse asyncResponse, boolean auth
-                    });
-                }
-            }
-            NamespaceBundles bundles = pulsar().getNamespaceService().getNamespaceBundleFactory()
-                    .getBundles(namespaceName);
-            for (NamespaceBundle bundle : bundles.getBundles()) {
-                // check if the bundle is owned by any broker, if not then we do not need to delete the bundle
-                if (pulsar().getNamespaceService().getOwner(bundle).isPresent()) {
-                    futures.add(pulsar().getAdminClient().namespaces()
-                            .deleteNamespaceBundleAsync(namespaceName.toString(), bundle.getBundleRange()));
+        // remove system topics first.
+        if (!topics.isEmpty()) {
+            for (String topic : topics) {
+                try {
+                    futures.add(pulsar().getAdminClient().topics().deleteAsync(topic, true, true));
+                } catch (Exception ex) {
+                    log.error("[{}] Failed to delete system topic {}", clientAppId(), topic, ex);
+                    asyncResponse.resume(new RestException(Status.INTERNAL_SERVER_ERROR, ex));
+                    return;
                 }
             }
-        } catch (Exception e) {
-            log.error("[{}] Failed to remove owned namespace {}", clientAppId(), namespaceName, e);
-            asyncResponse.resume(new RestException(e));
-            return;
         }
-
-        FutureUtil.waitForAll(futures).handle((result, exception) -> {
-            if (exception != null) {
-                if (exception.getCause() instanceof PulsarAdminException) {
-                    asyncResponse.resume(new RestException((PulsarAdminException) exception.getCause()));
-                    return null;
-                } else {
-                    log.error("[{}] Failed to remove owned namespace {}", clientAppId(), namespaceName, exception);
-                    asyncResponse.resume(new RestException(exception.getCause()));
-                    return null;
+        FutureUtil.waitForAll(futures).thenCompose(__ -> {
+            List<CompletableFuture<Void>> deleteBundleFutures = Lists.newArrayList();
+            try {
+                NamespaceBundles bundles = pulsar().getNamespaceService().getNamespaceBundleFactory()
+                        .getBundles(namespaceName);
+                for (NamespaceBundle bundle : bundles.getBundles()) {
+                    // check if the bundle is owned by any broker, if not then we do not need to delete the bundle
+                    if (pulsar().getNamespaceService().getOwner(bundle).isPresent()) {
+                        deleteBundleFutures.add(pulsar().getAdminClient().namespaces()
+                                .deleteNamespaceBundleAsync(namespaceName.toString(), bundle.getBundleRange()));
+                    }
                 }
+                return FutureUtil.waitForAll(deleteBundleFutures);
+            } catch (Exception e) {
```

Comment:
Why do we need this catch block?

Reply:
`pulsar().getNamespaceService().getOwner(bundle)` and `pulsar().getAdminClient()` has declared exception。

Reply:
@eolivelli

Reply:
> 

Thanks, @eolivelli, I get your point.

Reply:
Using getOwnerAsync instead of getOwner.